### PR TITLE
Migrate Python samples to uv and update the existing test script

### PR DESCRIPTION
**Pull request by datruong-ibmmq**
### Summary

This PR modernises the Python samples' dependency management by adopting [`uv`](https://docs.astral.sh/uv/) and minor update on the existing integration test script (`tests/test-python.sh`), which still use the plain venv. 

---

### Changes

- **Added [`Python/pyproject.toml`](Python/pyproject.toml)** — standard project manifest declaring `ibmmq` as the sole runtime dependency and pinning `requires-python = ">=3.9"`. Replaces ad-hoc `pip install` calls.
- **Added [`Python/uv.lock`](Python/uv.lock)** — deterministic lock file generated by `uv lock`. Pins `ibmmq 2.1.1` with verified hashes for macOS arm64, Linux x64, and Windows x64 wheels. Commit this file for reproducible installs.

#### Integration test script — [`tests/test-python.sh`](tests/test-python.sh)

- Quoted the variable $venv to support spaces
- Factorised /tmp/rc
- The script is functionally compatible with the updated Python files. ie. The six scripts run correctly as python <file>.py from the Python/ directory. 

**Notes:** The main long-term risk is the disconnect between pyproject.toml and the bare pip install ibmmq in the test script — careful maintenance whenever new dependencies are added.

#### Documentation

- **[`Python/README.md`](Python/README.md)** — Written with uv and venv paths; corrected typo `basicsubcribe` → `basicsubscribe` fixed.
- **[`README.md`](README.md)** — A typo correction at the very last line.

---

### How to verify

```bash
# Sync dependencies (requires uv and MQ C client libraries)
cd Python
uv sync

# Run the integration tests against a live queue manager
cd tests
bash test-python.sh
```

The script exits 0 when all 6 samples (`PUT`, `GET`, `PUB`, `SUB`, `REQ`, `RES`) complete successfully, and writes a full log to `tests/logs/testpython.log`.

### CODE REVIEW

File: Python/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
+# -*- coding: utf-8 -*-
+"""Shared pytest fixtures for the Python IBM MQ sample tests.
+
+The ibmmq C-extension is not available in CI/unit-test environments, so every
```
Comment (chughts) [inline Python/tests/conftest.py:4]: Does this mean that this module needs to be updated every time any of the structures are changed?

File: Python/TECHNICAL_SPEC.md [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+| `TestBasicGet` | 6 | `build_mq_details` (with index); `connect` returns QueueManager / None on error; `get_queue`; `get_messages` exits on empty queue; `get_messages` logs received message |
+| `TestBasicPublish` | 6 | `build_mq_details` includes `TOPIC_NAME`; `connect`; `get_topic`; `publish_message` appends, payload JSON, swallows error |
+| `TestBasicSubscribe` | 5 | `build_mq_details` includes `TOPIC_NAME`; `connect`; `get_subscription` (`SD.set_vs`); `get_messages` exits on empty; `get_messages` logs publication |
+| `TestBasicRequest` | 7 | `build_mq_details` includes MODEL/DYNAMIC keys; `connect`; `put_message` returns MsgId, sets `MQMT_REQUEST`, sets `ReplyToQ`; `await_response` exits on empty queue, logs reply |
+| `TestBasicResponse` | 14 | `perform_calc` (prime=[], 1=[], 12=[2,2,3], 60 divisible, 100 divisible); `rollback` backout path; `rollback` poison-message `AttributeError`; `rollback` returns False on `MQMIError`; `respond_to_request` calls `put1` returns True, sets `MQMT_REPLY`, copies MsgId to CorrelId, returns False on `put1` error; `get_messages` commits on success, exits on empty queue |
+
+### 9.3 Running the Tests
+
+```bash
```
Comment (chughts) [inline Python/TECHNICAL_SPEC.md:378]: Should we not be promoting the use of Python virtual environments eg. uv. ?

**Review by chughts — CHANGES_REQUESTED** [review]
Looks OK, but I think we should be promoting Python virtual environments eg. `uv` and consequently have a project.toml file to define the needed dependencies.

**Review by ibmmqmet — COMMENTED** [review]
I really dislike the TECHNICAL_SPEC file. Clearly written by an LLM, and it actually makes the whole thing look a lot more complicated than it is.

I also dislike mocking the qmgr/API. 

See the `tests/test-python` file under the root directory for simple tests of the Python samples that use a real qmgr.

File: Python/TECHNICAL_SPEC.md (hunk 2 above)
Comment (datruong-ibmmq) [reply Python/TECHNICAL_SPEC.md:378]: I will look into uv, and see how we would integrate/use it.
But, this means, another 3rdpty tool to be installed & use.

File: Python/tests/conftest.py (hunk 1 above)
Comment (datruong-ibmmq) [reply Python/tests/conftest.py]: These MQ stubs are setup for current unit tests. If there's signiticant change in future MQ and/or API, then update would be required.

**Comment by datruong-ibmmq** [comment]
> I really dislike the TECHNICAL_SPEC file. Clearly written by an LLM, and it actually makes the whole thing look a lot more complicated than it is.

DAT> I was thinking, a spec like this would help a new comer/student/user. I can discard or simplify it, if you prefer.

> I also dislike mocking the qmgr/API.

DAT> This is a mock/stub tests, not e2e.
 
> See the `tests/test-python` file under the root directory for simple tests of the Python samples that use a real qmgr.

DAT> OK, this can be expanded to perform more e2e test cases/scenarios, if needed. Also, this was merged via PR478, after this PR470.

**Comment by datruong-ibmmq** [comment]
@ibmmqmet - Mark, I've checked with explanation below to clarify the diferences and that they are complementary, not redundant. It does make sense to have both sets of tests., where the shell script answers *"does the full stack work end-to-end against a real broker?"*, while the Python tests answer *"does each function behave correctly under all conditions, including failures?"* 

---

## Overview

The two test suites test the same six Python MQ sample scripts (`basicput`, `basicget`, `basicpublish`, `basicsubscribe`, `basicrequest`, `basicresponse`) but are fundamentally different in design, execution environment, and granularity.

---

## [`tests/test-python.sh`](tests/test-python.sh) — Shell Integration Test

### What it does
A Bash integration test that runs the six sample scripts as **real OS processes** against a **live IBM MQ broker**. It:

1. Creates (or reuses) a Python `venv` and installs the real `ibmmq` package from PyPI
2. Launches `basicsubscribe.py` and `basicresponse.py` as background processes, then runs their counterparts (`basicpublish.py`, `basicrequest.py`) in the foreground
3. Counts the number of `"ended OK"` lines in the log — expects exactly **6 successes**

### Key characteristics
| Aspect | Detail |
|---|---|
| **Execution** | Spawns actual Python processes via `python script.py` |
| **MQ dependency** | Requires a **live, reachable MQ queue manager** with configured queues/topics |
| **Assertion style** | Binary pass/fail via process exit codes (`$?`) and log-line counting |
| **Concurrency** | Uses `&` / `sleep 1` / `wait` to coordinate pub/sub and req/resp pairs |
| **Scope** | End-to-end: the full script runs top to bottom, including `connect()`, `put()`, `get()`, `disconnect()` |
| **Granularity** | One assertion per script — does it exit 0 or not? |
| **Speed** | Slow — real network I/O, `sleep 1` delays, pip install |
| **Portability** | Requires Bash, a running MQ broker, and the native `ibmmq` C extension |

---

## [`Python/tests/`](Python/tests/) — Python Unit Tests (pytest)

Three files work together:

### [`conftest.py`](Python/tests/conftest.py) — Shared fixtures
Provides two fixtures available to all tests:
- **`stub_ibmmq`** (`autouse=True`) — builds a complete in-process fake of the `ibmmq` C extension. Stubs out `QueueManager`, `Queue`, `Topic`, `Subscription`, `MD`, `PMO`, `SD`, `CSP`, `MQMIError`, and all `CMQC`/`CMQXC` constants. No C extension or broker needed.
- **`env_json_file`** — writes a minimal `env.json` to a `tmp_path` and sets `JSON_CONFIG`, so `EnvStore` can load without any real environment setup.

### [`test_env.py`](Python/tests/test_env.py) — Unit tests for `utils/env.py`
Tests `EnvStore` in isolation across 6 classes (~25 test cases):
- Config loading, missing file, invalid JSON
- `is_endpoint_list()`, `build_connection_string()`, `get_endpoint_count()`
- `get_next_connection_string()`, `set_env()`, `getenv_value()`, `get_connection()`, `is_ccdt_available()`

### [`test_samples.py`](Python/tests/test_samples.py) — Unit tests for the 6 sample scripts
Tests individual **functions** within each sample module across 6 test classes (~40 test cases):
- Imports each module against the stub (which runs module-level code harmlessly)
- Directly calls `build_mq_details()`, `connect()`, `get_queue()`, `put_message()`, `get_messages()`, etc.
- Uses `monkeypatch` to inject failures and verify error-handling behaviour
- Asserts on internal state (`stub_queue._messages`, `qmgr._committed`, captured `md.MsgType`, etc.)

### Key characteristics
| Aspect | Detail |
|---|---|
| **Execution** | In-process, pure Python, no subprocesses |
| **MQ dependency** | **None** — the `ibmmq` C extension is fully stubbed |
| **Assertion style** | Fine-grained `assert` statements on return values, state, log output |
| **Scope** | Unit-level: individual functions, not full scripts |
| **Granularity** | ~40+ focused test cases covering happy paths, edge cases, and error paths |
| **Speed** | Fast — milliseconds, no I/O |
| **Portability** | Runs anywhere with Python 3 + pytest, no broker required |

---

## Side-by-side Summary

| Dimension | `test-python.sh` | `Python/tests/*.py` |
|---|---|---|
| **Type** | Integration / E2E | Unit |
| **Broker required** | ✅ Yes | ❌ No |
| **ibmmq C extension** | ✅ Real | ❌ Stubbed |
| **Assertion granularity** | Exit codes only | Return values, state, logs |
| **Error-path coverage** | None | Explicit (monkeypatch) |
| **Concurrency** | Bash `&` / `wait` | Not needed |
| **Run time** | Slow (real I/O + sleep) | Fast |
| **CI suitability** | Needs MQ infrastructure | Runs in any CI |
| **Samples covered** | All 6 (as processes) | All 6 (as imported modules) |
| **`env.py` coverage** | Implicit (used by scripts) | Explicit, dedicated test class |

**Comment by datruong-ibmmq** [comment]
@chughts - Soheel, the `uv` integration lives entirely in [`pyproject.toml`](Python/pyproject.toml), [`uv.lock`](Python/uv.lock), and the updated shell script/docs. The unit test files added here (Python/tests/*.py) are agnostic to how the venv is managed. But, it does affect `tests/test-python.sh`. Thus, I would defer `uv` integration to another (later) PR.

---

### [`conftest.py`](Python/tests/conftest.py) — No changes needed

It only uses Python stdlib (`json`, `os`, `sys`, `types`) and `pytest`. No path manipulation, no pip/venv references. `uv run pytest` places the managed venv on `sys.path` automatically, so `import pytest` just works.

---

### [`test_env.py`](Python/tests/test_env.py) — No changes needed

Only uses stdlib + `pytest`. The `_reset_env_store()` helper manipulates `sys.modules` directly — purely runtime behaviour, unaffected by which tool manages the venv.

---

### [`test_samples.py`](Python/tests/test_samples.py) — No changes needed, but worth understanding why the `sys.path` line is fine

The one line that looks like it might be fragile is in every `setup` fixture:

```python
_sys.path.insert(0, "Python")
```

This works because `uv run pytest` is invoked **from the `Python/` directory**, so the inserted path `"Python"` resolves to `Python/Python/` — which doesn't exist, but it doesn't matter. The sample modules (`basicput`, `basicget`, etc.) are found because `pytest` already adds the project root (`Python/`) to `sys.path` via its own import machinery when it collects tests. The `sys.path.insert` is a belt-and-braces no-op in this layout.

If you ever ran pytest from the repo root instead (`uv run pytest Python/tests/`), that line would correctly add `Python/` to the path so the sample imports resolve. Either way, no change needed.

---

### Summary

| File | Change needed? | Reason |
|---|---|---|
| [`conftest.py`](Python/tests/conftest.py) | ❌ No | Pure stdlib + pytest; no path or tooling assumptions |
| [`test_env.py`](Python/tests/test_env.py) | ❌ No | Same |
| [`test_samples.py`](Python/tests/test_samples.py) | ❌ No | `sys.path.insert(0, "Python")` is harmless and already correct for both invocation styles |

File: Python/TECHNICAL_SPEC.md (hunk 2 above)
Comment (datruong-ibmmq) [reply Python/TECHNICAL_SPEC.md:365]: The use of `uv` for Python isn't applicable to these unit tests. Relevant changes would be updated then.
It would be best to introduce `uv` under a different PR. See comments below for details

**Comment by datruong-ibmmq** [comment]
@chughts @ibmmqmet - I have removed the unit tests, as per today's conversation with @chughts. Thus, update this PR for the purpose of using `uv`, rather than unit testing in Python directory. See the updated description of this PR for details.

File: Python/tests/conftest.py (hunk 1 above)
Comment (datruong-ibmmq) [reply Python/tests/conftest.py]: File has been removed

File: Python/README.md [hunk 3]
```diff
@@ -64,28 +83,44 @@ On some systems, you might need to explicitly use the `python3` command instead
 ### Put/Get
 The `basicput` application places a short string message onto the queue.
 
-`python ./basicput`
+```bash
+uv run python basicput.py
```
Comment (chughts) [inline Python/README.md:87]: Only need
`uv run basicput.py`

This applies to all instances of `uv run` in this readme.

File: Python/README.md [hunk 4]
```diff
@@ -64,28 +83,44 @@ On some systems, you might need to explicitly use the `python3` command instead
...
 
-`python ./basicresponse`
+```bash
+uv run python basicresponse.py    # window 1
+uv run python basicrequest.py     # window 2
+```
+
+## Running the unit tests
+
+The tests require no live IBM MQ broker — the `ibmmq` C extension is fully stubbed.
```
Comment (chughts) [inline Python/README.md:120]: I thought the stubs had been removed.

File: Python/TECHNICAL_SPEC.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
+# Python IBM MQ Dev Patterns — Technical Specification
+
```
Comment (chughts) [inline Python/TECHNICAL_SPEC.md:2]: I got the impression that @ibmmqmet doesn't like this file.

File: tests/test-python.sh [hunk 6]
```diff
@@ -15,54 +15,45 @@ function checkRc {
...
-if [ ! -d $venv ]
-then
-  python -m venv venv
-  if [ $? -ne 0 ]
-  then
-    echo "ERROR: Cannot create Python virtual env at $venv"
-    exit 1
-  fi
+# Require uv — install it if absent (curl pipe to sh is the official installer)
+if ! command -v uv &>/dev/null; then
```
Comment (chughts) [inline tests/test-python.sh:19]: will not work for windows. 
https://docs.astral.sh/uv/getting-started/installation/#standalone-installer
says

`pip install uv`

File: Python/pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+[project]
+name = "mq-python-samples"
+version = "1.0.0"
+description = "IBM MQ Python sample applications demonstrating core messaging patterns"
+requires-python = ">=3.9"
+
+# Runtime dependency — the ibmmq C extension that wraps the MQI
+dependencies = [
+    "ibmmq",
```
Comment (chughts) [inline Python/pyproject.toml:9]: As we don't specify a version here, we need to be clear on how to install the current latest. Using the included lock file will only get the version in the lock file.

**Review by chughts — CHANGES_REQUESTED** [review]
Not quite cleaned up from the previous PR.

**Comment by ibmmqmet** [comment]
I also don't like uv. It requires installing something extra, wheras `venv` is part of the Python basic install. While it might have some advantages, for simple scenarios it's not needed. And `venv` also matches the main Python scripts.

**Comment by datruong-ibmmq** [comment]
> I also don't like uv. It requires installing something extra, wheras `venv` is part of the Python basic install. While it might have some advantages, for simple scenarios it's not needed. And `venv` also matches the main Python scripts.

The Python/README.md has been updated to support both `uv` and `venv`, ie. the dual-track would give developers a choice of local workflow. The `tests/test-python.sh` script has been reverted to use `venv`, rather than `uv`.

File: Python/pyproject.toml (hunk 7 above)
Comment (datruong-ibmmq) [reply Python/pyproject.toml:12]: Added comments as requested

**Comment by ibmmqmet** [comment]
`basicget` simply does not work: 
```
./basicget.py:47:33: F821 undefined name 'conn_info'
```

Moving code into a `main` function has also caused a large amount of linting errors, with pylint and flake8:

```
************* Module Python.basicget
basicget.py:47:32: E0602: Undefined variable 'conn_info' (undefined-variable)
basicget.py:150:0: C0116: Missing function or method docstring (missing-function-docstring)
basicget.py:151:4: W0601: Global variable 'MQDetails' undefined at the module level (global-variable-undefined)
basicget.py:151:4: W0601: Global variable 'qmgr' undefined at the module level (global-variable-undefined)
basicget.py:151:4: W0601: Global variable 'queue' undefined at the module level (global-variable-undefined)
basicget.py:156:4: C0103: Variable name "envStore" doesn't conform to snake_case naming style (invalid-name)
basicget.py:164:4: C0103: Variable name "numEndPoints" doesn't conform to snake_case naming style (invalid-name)
************* Module Python.basicpublish
basicpublish.py:106:0: C0116: Missing function or method docstring (missing-function-docstring)
basicpublish.py:107:4: W0601: Global variable 'MQDetails' undefined at the module level (global-variable-undefined)
basicpublish.py:107:4: W0601: Global variable 'conn_info' undefined at the module level (global-variable-undefined)
basicpublish.py:107:4: W0601: Global variable 'msg_object' undefined at the module level (global-variable-undefined)
basicpublish.py:107:4: W0601: Global variable 'qmgr' undefined at the module level (global-variable-undefined)
basicpublish.py:107:4: W0601: Global variable 'topic' undefined at the module level (global-variable-undefined)
basicpublish.py:112:4: C0103: Variable name "envStore" doesn't conform to snake_case naming style (invalid-name)
************* Module Python.basicput
basicput.py:112:0: C0116: Missing function or method docstring (missing-function-docstring)
basicput.py:113:4: W0601: Global variable 'MQDetails' undefined at the module level (global-variable-undefined)
basicput.py:113:4: W0601: Global variable 'conn_info' undefined at the module level (global-variable-undefined)
basicput.py:113:4: W0601: Global variable 'msg_object' undefined at the module level (global-variable-undefined)
basicput.py:113:4: W0601: Global variable 'qmgr' undefined at the module level (global-variable-undefined)
basicput.py:113:4: W0601: Global variable 'queue' undefined at the module level (global-variable-undefined)
basicput.py:118:4: C0103: Variable name "envStore" doesn't conform to snake_case naming style (invalid-name)
************* Module Python.basicrequest
basicrequest.py:208:0: C0116: Missing function or method docstring (missing-function-docstring)
basicrequest.py:209:4: W0601: Global variable 'MQDetails' undefined at the module level (global-variable-undefined)
basicrequest.py:209:4: W0601: Global variable 'conn_info' undefined at the module level (global-variable-undefined)
basicrequest.py:209:4: W0601: Global variable 'msg_object' undefined at the module level (global-variable-undefined)
basicrequest.py:209:4: W0601: Global variable 'qmgr' undefined at the module level (global-variable-undefined)
basicrequest.py:209:4: W0601: Global variable 'queue' undefined at the module level (global-variable-undefined)
basicrequest.py:209:4: W0601: Global variable 'dynamic' undefined at the module level (global-variable-undefined)
basicrequest.py:214:4: C0103: Variable name "envStore" doesn't conform to snake_case naming style (invalid-name)
************* Module Python.basicresponse
basicresponse.py:287:0: C0116: Missing function or method docstring (missing-function-docstring)
basicresponse.py:288:4: W0601: Global variable 'MQDetails' undefined at the module level (global-variable-undefined)
basicresponse.py:288:4: W0601: Global variable 'conn_info' undefined at the module level (global-variable-undefined)
basicresponse.py:288:4: W0601: Global variable 'qmgr' undefined at the module level (global-variable-undefined)
basicresponse.py:288:4: W0601: Global variable 'queue' undefined at the module level (global-variable-undefined)
basicresponse.py:293:4: C0103: Variable name "envStore" doesn't conform to snake_case naming style (invalid-name)
************* Module Python.basicsubscribe
basicsubscribe.py:148:0: C0116: Missing function or method docstring (missing-function-docstring)
basicsubscribe.py:149:4: W0601: Global variable 'MQDetails' undefined at the module level (global-variable-undefined)
basicsubscribe.py:149:4: W0601: Global variable 'conn_info' undefined at the module level (global-variable-undefined)
basicsubscribe.py:149:4: W0601: Global variable 'qmgr' undefined at the module level (global-variable-undefined)
basicsubscribe.py:149:4: W0601: Global variable 'subscription' undefined at the module level (global-variable-undefined)
basicsubscribe.py:154:4: C0103: Variable name "envStore" doesn't conform to snake_case naming style (invalid-name)

```

Code should always be linted and tested before submission.
